File: data_structures/trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_string = False
    
    def __repr__(self):
        return f'{self.children}, {self.end_of_string}'
    
    def __bool__(self):
        return bool(self.children)

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str):
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.end_of_string = True
    
    def search(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        return current.end_of_string

def deleteString(root: TrieNode, word, index=0):
    char = word[index]
    current_node = root.children.get(char)
    delete = False
    if len(current_node.children) > 1:
        deleteString(current_node, word, index + 1)
        return False
    if index == len(word) - 1:
        if len(current_node.children) >= 1:
            current_node.end_of_string = False
            return False
        else:
            root.children.pop(char)
            return True
    if current_node.end_of_string == True:
        deleteString(current_node, word, index + 1)
        return False
    delete = deleteString(current_node, word, index + 1)
    if delete is True:
        root.children.pop(char)
        return True
    else:
        return False

File: go_files/system_design/data_stream_processing/trie.go
package stream_processing

import "fmt"

type TrieNode struct {
	Children map[rune]*TrieNode
	End      bool
}

func NewTrieNode() *TrieNode {
	return &TrieNode{make(map[rune]*TrieNode, 26), false}
}

func (t *TrieNode) GoString() string {
	return fmt.Sprintf("TrieNode(Children: %#v, End: %v)", t.Children, t.End)
}

type Trie struct {
	Root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{NewTrieNode()}
}

func (t *Trie) InsertReverse(word string) {
	current := t.Root
	runes := []rune(word)
	for i := len(runes) - 1; i >= 0; i-- {
		currentRune := runes[i]
		if _, ok := current.Children[currentRune]; !ok {
			current.Children[currentRune] = NewTrieNode()
		}
		current = current.Children[currentRune]
	}
	current.End = true
}

func (t *Trie) Search(word []rune) bool {
	current := t.Root

	for i := len(word) - 1; i >= 0; i-- {
		currenNode, ok := current.Children[word[i]]
		if !ok {
			return false
		}
		current = currenNode
		if current.End {
			return true
		}
	}
	return false
}
